src/components/menu/positioning/positioningCalculations.ts

```diff
@@ -0,0 +1,209 @@
+    left: number;
+    top: number;
+    right: number;
+    bottom: number;
+}
+
+export interface IDimensionsAndCoordinates extends IDimensions, ICoordinates {}
+
+export type Dimension = 'width' | 'height';
+
+export type Direction = 'left' | 'right' | 'top' | 'bottom';
+
+export interface IMenuPosition {
+    left: number;
+    top: number;
+}
+
+export function getViewportDimensionsAndCoords(): IDimensionsAndCoordinates {
+    const width = window.innerWidth;
+    const height = window.innerHeight;
+
+    const left = window.pageXOffset;
+    const top = window.pageYOffset;
+    const right = left + width;
+    const bottom = top + height;
+
+    return { width, height, left, top, right, bottom };
+}
+
+export function getElementDimensionsAndCoords(element: HTMLElement): IDimensionsAndCoordinates {
+    const rect = element.getBoundingClientRect();
+    const window = getViewportDimensionsAndCoords();
+
+    const width = rect.width;
+    const height = rect.height;
+
+    const left = rect.left + window.left;
+    const top = rect.top + window.top;
+    const right = left + width;
+    const bottom = top + height;
+
+    return { width, height, left, top, right, bottom };
+}
+
+export function getElementDimensions(element: Element): IDimensions {
+    const rect = element.getBoundingClientRect();
+    const width = rect.width;
+    const height = rect.height;
+
+    return { width, height };
+}
+
+const reverseDirectionMap: { [key in Direction]: Direction } = {
+    left: 'right',
+    right: 'left',
+    top: 'bottom',
+    bottom: 'top'
+};
+
+const dimensionMap: { [key in Direction]: Dimension } = {
+    left: 'width',
+    right: 'width',
+    top: 'height',
+    bottom: 'height'
+};
+
+export interface IMenuPositionConfig {
+    toggler: IDimensionsAndCoordinates;
+    viewport: IDimensionsAndCoordinates;
+    menu: IDimensions;
+    alignment: MenuAlignment;
+    spacing: number;
+    offset: number;
+    topLevelMenu: boolean;
+}
+
+export function calculateMenuPosition({
+    toggler,
+    viewport,
+    menu,
+    alignment = ['right', 'bottom'],
+    spacing = 0,
+    offset = 0,
+    topLevelMenu = true
+}: IMenuPositionConfig): IMenuPosition {
+    const sharedArguments = { toggler, viewport, menu, spacing, offset };
+    const [directionPreferredPrimary, directionPrefferedSecondary] = alignment;
+
+    const [primaryCoordinateDirection, primaryCoordinate] = calculatePositionForDirection({
+        ...sharedArguments,
+        direction: directionPreferredPrimary,
+        isPrimaryDimension: true
+    });
+    const [secondaryCoordinateDirection, secondaryCoordinate] = calculatePositionForDirection({
+        ...sharedArguments,
+        direction: directionPrefferedSecondary,
+        isPrimaryDimension: false
+    });
+
+    const coordinates: Partial<ICoordinates> = {
+        [primaryCoordinateDirection]: primaryCoordinate,
+        [secondaryCoordinateDirection]: secondaryCoordinate
+    };
```

Comment:
Alternative way to write this without `const [] = ...` if calculatePositionForDirection returned `{ left: 123 }` instead `['left', 123]`;

```
    const primaryCoordinate = calculatePositionForDirection({
        ...sharedArguments,
        direction: directionPreferredPrimary,
        isPrimaryDimension: true
    });
    const secondaryCoordinate = calculatePositionForDirection({
        ...sharedArguments,
        direction: directionPrefferedSecondary,
        isPrimaryDimension: false
    });

const coordinates = {
    ... primaryCoordinate,
    ... secondaryCoordinate
}
```